UIKit/Location Controller/TWTLocationController.m

```diff
@@ -0,0 +1,189 @@
+        [self stopUpdatingLocation];
+    }
+    else {
+        NSLog(@"location timed out.");
+        [self failWithErrorCode:TWTLocationControllerErrorUpdateTimedOut underlyingError:nil];
+    }
+}
+
+
+#pragma mark - CLLocationManagerDelegate
+
+- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
+{
+    [self updateLocationForAuthorizationStatus:status];
+}
+
+
+- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations
+{
+    self.location = [locations lastObject];
+
+    if ([self validateLocation:self.location accuracy:manager.desiredAccuracy]) {
+        [self processCompletionBlocksWithLocation:self.location error:nil];
+        [self stopUpdatingLocation];
+    }
+}
+
+
+- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
+{
+    // TODO (foster): What to pass here as an error code.
+    [self failWithErrorCode:error.code underlyingError:error];
+}
+
+
+#pragma mark - Helpers
+
+- (void)updateLocationForAuthorizationStatus:(CLAuthorizationStatus)status
+{
+    switch ([CLLocationManager authorizationStatus]) {
+
+        case kCLAuthorizationStatusNotDetermined: {
+            [self failWithErrorCode:TWTLocationControllerAuthorizationStatusNotDetermined underlyingError:nil];
+            [self.locationManager requestWhenInUseAuthorization];
+            [self.locationManager requestAlwaysAuthorization];
+            break;
+        }
+
+        case kCLAuthorizationStatusRestricted: {
+            [self failWithErrorCode:TWTLocationControllerAuthorizationStatusRestricted underlyingError:nil];
+            break;
+        }
+
+        case kCLAuthorizationStatusDenied: {
+            [self failWithErrorCode:TWTLocationControllerErrorPermissionDenied underlyingError:nil];
+            break;
+        }
+
+        case kCLAuthorizationStatusAuthorizedWhenInUse:
+        case kCLAuthorizationStatusAuthorizedAlways: {
+
+            if ([self validateLocation:self.location accuracy:self.locationManager.desiredAccuracy]) {
+                [self processCompletionBlocksWithLocation:self.location error:nil];
+            }
+            else {
+                [self startUpdatingLocation];
+            }
+            break;
+        }
+    }
+}
+
+
+- (void)failWithErrorCode:(TWTLocationControllerStatus)code underlyingError:(NSError *)underlyingError
+{
+    NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
+    if (underlyingError) {
+        userInfo[NSUnderlyingErrorKey] = underlyingError;
+    }
+
+    NSError *error = [NSError errorWithDomain:kLocationControllerErrorDomain code:code userInfo:userInfo];
+    [self processCompletionBlocksWithLocation:nil error:error];
+}
+
+
+- (void)startUpdatingLocation
+{
+    [self.locationManager startUpdatingLocation];
+}
+
+
+- (void)stopUpdatingLocation
+{
+    [self.locationManager stopUpdatingLocation];
+}
```

Comment:
Not entirely clear what the value of these methods are. Do you expect them to grow?